modules/calcite/src/test/java/org/apache/ignite/internal/processors/query/calcite/integration/SqlPlanHistoryIntegrationTest.java

```diff
@@ -291,141 +288,168 @@ public void testJdbcQueryFailed() throws Exception {
     }
 
     /** Checks successful cross-cache SqlFieldsQuery. */
     @Test
-    public void testSqlFieldsCrossCacheQuery() {
+    public void testSqlFieldsCrossCacheQuery() throws Exception {
         runSuccessfulQuery(sqlFieldsQryCrossCache);
     }
 
     /** Checks failed cross-cache SqlFieldsQuery. */
     @Test
-    public void testSqlFieldsCrossCacheQueryFailed() {
+    public void testSqlFieldsCrossCacheQueryFailed() throws Exception {
         runFailedQuery(sqlFieldsQryCrossCacheFailed);
     }
 
-    /** Checks successful SqlFieldsQuery with reduce phase. */
+    /**
+     * Checks successful SqlFieldsQuery with reduce phase. There should be 3 entries in SQL plan history on the
+     * reduce node (2 SELECT subqueries and 1 UNION/MERGE subquery) and 2 entries on the map node (2 SELECT subqueries).
+     */
     @Test
-    public void testSqlFieldsQueryWithReducePhase() {
+    public void testSqlFieldsQueryWithReducePhase() throws Exception {
         runQueryWithReducePhase(() -> {
-            cacheQuery(sqlFieldsQryWithReducePhase.setLocal(loc), "pers");
+            try {
+                startGridsMultiThreaded(1, 2);
+
+                awaitPartitionMapExchange();
+            }
+            catch (Exception e) {
+                throw new RuntimeException(e);
+            }
+
+            cacheQuery(sqlFieldsQryWithReducePhase, "pers");
+
+            checkSqlPlanHistory(3);
 
-            checkSqlPlanHistory((!isClient && sqlEngine == IndexingQueryEngineConfiguration.ENGINE_NAME) ? 3 : 1);
+            for (int i = 1; i <= 2; i++) {
+                Map<SqlPlan, Long> sqlPlansOnMapNode = grid(i).context().query().runningQueryManager()
+                    .planHistoryTracker().sqlPlanHistory();
+
+                assertNotNull(sqlPlansOnMapNode);
+
+                checkMetrics(2, sqlPlansOnMapNode);
+            }
         });
     }
 
-    /** Checks failed SqlFieldsQuery with reduce phase. */
+    /**
+     * Checks failed SqlFieldsQuery with reduce phase. If the fisrt subquery fails, there should be only one entry in
+     * SQL plan history. If the fisrt subquery is successfully executed but the second one fails, the history should
+     * contain two entries.
+     */
     @Test
-    public void testSqlFieldsQueryWithReducePhaseFailed() {
+    public void testSqlFieldsQueryWithReducePhaseFailed() throws Exception {
         runQueryWithReducePhase(() -> {
-            try {
-                cacheQuery(sqlFieldsQryWithReducePhaseFailed.setLocal(loc), "pers");
-            }
-            catch (Exception ignore) {
-                //No-Op
-            }
+            for (int i = 0; i < sqlFieldsQryWithReducePhaseFailed.length; i++) {
+                try {
+                    cacheQuery(sqlFieldsQryWithReducePhaseFailed[i], "pers");
+                }
+                catch (Exception ignore) {
+                    // No-op.
+                }
+
+                checkSqlPlanHistory( i + 1);
```

Comment:
It's not intuitive clear that map-subqueries are executed one by one. It's not stated anywhere and subqueries can be run in parallel with some tuning or after some change in future. So, perhaps we should remove this test at all, since can't assert correct behaviour.

Reply:
Test deleted.